### ISSUE TITLE
Scheduler prepares work on same worker, while there's others available

### ISSUE THREAD

**Issue by RobQuistNL** [filecoin-project/lotus#8655]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] This is **not** a question or a support request. If you have any lotus related questions, please ask in the [lotus forum](https://github.com/filecoin-project/lotus/discussions).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** an enhancement request. If it is, please file a [improvement suggestion](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fenhancement&template=enhancement.yml) instead.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [ ] I did not make any code changes to lotus.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [X] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [ ] lotus miner/market - storage deal
- [ ] lotus miner/market - retrieval deal
- [ ] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [ ] Other

### Lotus Version

```text
v1.15.3-rc1
```

### Describe the Bug

I have PC2 machines available, but the scheduler throws a PC2 job in "prepared" if they happen at the same time.

I run 2 pledges at once, PC1 finishes pretty much at the same time, and then;

```
88c9ce67  915     8d7a5f9b  pc2-04-pc2-2  PC2   running   13m51.1s
00000000  913     8d7a5f9b  pc2-04-pc2-2  PC2   prepared  13m51.1s
```

Makes no sense, because there's about 20 other PC2 workers available which are doing nothing.

### Logging Information

```text
-
```

### Repo Steps

1. Run `lotus-miner sectors pledge` a couple of times, quickly
2. Notice that a few workers have "prepared" PC2 jobs
3. Notice that a few workers are doing nothing
4. Notice that the times here are almost identical - pointing towards the "windowing" logic.

If we would just have a simple round-robin scheduler, we can manage the rest.

We know how much a machine can handle. We know how fast to pledge / seal.. We'd just like all the machines to be used, one by one, regardless of "windows" or resource usage.

**Comment by RobQuistNL** [comment]
Related: https://github.com/filecoin-project/lotus/issues/7868
https://github.com/filecoin-project/lotus/issues/7607
https://github.com/filecoin-project/lotus/issues/3709

https://github.com/filecoin-project/lotus/pull/8447

**Comment by RobQuistNL** [comment]
Still happens and is easy to reproduce. I think it happens when a job ends up in the same window.

I have 80 PC2 workers, yet;

```
ID        Sector  Worker    Hostname   Task  State     Time
e42b033a  0       d4536139  pc2-02     PC2   running   17m38.7s
3cbc2585  2       ea32b44b  pc2-02     PC2   running   15m4.6s
057b2937  4       5173c5c2  pc2-02     PC2   running   14m39.2s
e6660bd7  3       97963713  pc2-02     PC2   running   13m53.1s
c90ab599  1       988f7438  pc2-01     PC2   running   13m32.6s
417254d3  7       fbc4dff3  pc2-01     PC2   running   11m50.9s
02467457  5       dca9ffa3  pc2-01     PC2   running   11m28.9s
fcfa6ae8  9       b670329d  pc2-01     PC2   running   6m12s
481e1394  8       3d4f6523  pc2-11     PC2   running   5m6.7s
60b2e77d  10      53edcccc  pc2-04     PC2   running   4m28.8s
00000000  6       fbc4dff3  pc2-01     PC2   prepared  11m50.8s
```

see how 417254d3 and 00000000 have identical times?

So, to reproduce, make sure you finish 2 PC1 jobs at the same time. Then this happens.

**Comment by rjan90** [comment]
Hey @RobQuistNL! I think you where planning to cherry-pick and test out #8700? Would be really nice if you could give an update if the new `spread` logic option helps with spreading the work across all workers better, in your case here, across all your PC2 workers. Else I will try to test it out later this week / early next week. 😄 

The new logic should:
```
Introduce new, simpler worker assigning logic - when enabled, the "spread" assigner will ignore worker utilization, and will attempt to assign tasks to as many workers as possible

[Storage]
 # Assigner specifies the worker assigner to use when scheduling tasks.
 # "utilization" (default) - assign tasks to workers with lowest utilization.
 # "spread" - assign tasks to as many distinct workers as possible.
 #
 # type: string
 # env var: LOTUS_STORAGE_ASSIGNER
 Assigner = "spread"
```

**Comment by RobQuistNL** [comment]
Yep, already thanked Magic for his work and was about to test this :)

**Comment by hyunmoon** [comment]
I just started sealing with this code as well.
I'm glad I sticked around enough to see this day coming.

**Comment by hyunmoon** [comment]
I just started sealing with this code as well.
I'm glad I sticked around long enough to see this day coming.

**Comment by hyunmoon** [comment]
Not looking good. I'm getting many PC2 tasks in 'prepared' state while the workers that the sectors have been assigned onto are not performing any PC2 or GET.
I made sure the workers have enough resource available and of course PC2 tasks is enabled.
.. Why?
The cache files are not being transferred, and the worker is only doing a few PC1 (it can do a bunch of it).
Nothing's happening, sectors just stack up in prepared state.

**Comment by hyunmoon** [comment]
There's definitely a problem in SDR proofs code btw.
Some sectors finish PC1 unusually fast and they end up having to be removed because they cause rust panic no unwinding info error.
Multicore SDR disabled.

**Comment by hyunmoon** [comment]
I might as well roll back to v1.15.0 really

**Comment by hyunmoon** [comment]
Unfortunately I couldn't really notice any improvements in terms of spreading out the tasks evenly.
This issue seems to be an inherent issue of window based scheduling, not the resource utilization vs. spread-out algorithm thing.

**Comment by hyunmoon** [comment]
The scheduling table will have to be refreshed periodically to find best available worker in real time.
Not look up once, queue it, and forget about it strategy.

**Comment by hyunmoon** [comment]
For example, there's no need for a worker to have a 'prepared' and multiple 'assigned' C2 tasks while there are a bunch of workers idling.

**Comment by hyunmoon** [comment]
Give a call ID to each assigned tasks so they can be aborted manually?
Hopefully it gets re-assigned to a currently available worker.
